### PR TITLE
nixos/scion: init

### PR DESCRIPTION
**Pull request by MatthewCroughan**
## Description of changes

Adds nixosModules and an extensive VM Test for the [SCION internet architecture](https://scion-architecture.net/) which implements a 5 autonomous system network topology [based upon upstream documentation](https://github.com/scionproto/scion/blob/27983125bccac6b84d1f96f406853aab0e460405/doc/tutorials/deploy.rst)

Thanks to @roberth for help with figuring out how to structure the modules and their naming.

#### VM test topology

<a href="https://github.com/scionproto/scion/blob/27983125bccac6b84d1f96f406853aab0e460405/doc/tutorials/deploy.rst#sample-scion-demo-topology">
  
</a>

#### Goals for a future PR, **not this PR**

- module options for bootstrapping PKI similar to how openssh may generate `/etc/ssh/ssh_host*`
- module options for connecting immediately to https://www.scionlab.org/ (probably a sensible default for `scion.enable = true`)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/sc/scion/package.nix [hunk 1]
```diff
@@ -41,6 +41,13 @@ buildGo121Module {
     ${skipTest 1 "TestOpensslCompatible" "scion-pki/trcs/sign_test.go"}
   '';
 
+  postInstall = ''
+    ln -s $out/bin/dispatcher $out/bin/scion-dispatcher
+    ln -s $out/bin/router $out/bin/scion-router
+    ln -s $out/bin/control $out/bin/scion-control
+    ln -s $out/bin/daemon $out/bin/scion-daemon
```
Comment (roberth) [inline pkgs/by-name/sc/scion/package.nix:48]: These original names look way too generic. Are they even meant to be put onto `PATH`? Should we hide them?

**Review by roberth — COMMENTED** [review]
I know almost nothing about SCION, so this is not really a review, but from what I can tell the Nix/NixOS sauce looks good. (Also @MatthewCroughan is giving me way too much credit)

File: pkgs/by-name/sc/scion/package.nix (hunk 1 above)
Comment (MatthewCroughan) [reply pkgs/by-name/sc/scion/package.nix:48]: Yeah, good call, we should simply remove them in favor of the less generic names, to make sure nothing has a problem in the future.

**Comment by roberth** [comment]
> #### VM test topology
> [ ](https://github.com/scionproto/scion/blob/27983125bccac6b84d1f96f406853aab0e460405/doc/tutorials/deploy.rst#sample-scion-demo-topology)

Could you add this to a README in the test directory?

**Comment by MatthewCroughan** [comment]
@roberth Done, I've added it in `nixos/tests/scion/freestanding-deployment/README.rst`

File: nixos/modules/services/networking/scion/scion-dispatcher.nix [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    system.activationScripts.scion-dispatcher = ''
+      ln -sf /dev/shm /run/shm
+    '';
+
+    systemd.services.scion-dispatcher = {
+      description = "SCION Dispatcher";
+      after = [ "network-online.target" ];
+      wants = [ "network-online.target" ];
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
```
Comment (sarcasticadmin) [inline nixos/modules/services/networking/scion/scion-dispatcher.nix:61]: The following `ExecStartPre` might be needed for the dispatcher to cleanup old files: 
```
ExecStartPre=/bin/rm -rf /run/shm/dispatcher/
```
> Ref: https://github.com/netsec-ethz/scion-builder/blob/7ee515d2419d6f2f35cec76d83fa33f460ec357a/systemd/scion-dispatcher.service#L12

Im not sure how critical this is but at least wanted to mention it for discussion

**Review by sarcasticadmin — COMMENTED** [review]
@MatthewCroughan awesome job getting this push up! Its exciting to have the SCION modules and this freestanding test.

Had a couple of comments but overall I think this looks solid. I mentioned this for the control daemon but for all the daemons included here: Would it be possible to include a `environment.etc."scion/<daemon>.conf".source = confFile;`? Its really nice to have the unit files reference things in `/etc` instead of `/nix/store` :slightly_smiling_face: 

Running `nix build .#nixosTests.scion-freestanding-deployment` passes with flying colors :rocket:

File: nixos/modules/services/networking/scion/scion-control.nix [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  config = mkIf cfg.enable {
+    systemd.services.scion-control = {
+      description = "SCION Control Service";
+      after = [ "network-online.target" "scion-dispatcher.service" ];
+      wants = [ "network-online.target" "scion-dispatcher.service" ];
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
+        Type = "simple";
+        Group = if (config.services.scion.scion-dispatcher.enable == true) then "scion" else null;
+        ExecStart = "${pkgs.scion}/bin/scion-control --config ${configFile}";
```
[the remark this answers is not in the document]
Comment (MatthewCroughan) [reply nixos/modules/services/networking/scion/scion-control.nix:61]: If we now want to put it in /etc, we need to give it a folder and a name, which doesn't mean anything in the context of how we're using it (just passing arguments via an attrset).

File: nixos/tests/scion/freestanding-deployment/default.nix [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+      networking = {
+        useNetworkd = true;
+        useDHCP = false;
+      };
+      systemd.network.networks."01-eth1" = {
+        name = "eth1";
+        networkConfig.Address = "192.168.1.${toString hostId}/24";
+      };
+      environment.etc = {
+        "scion/topology.json".source = ./topology${toString hostId}.json;
```
[the remark this answers is not in the document]
Comment (MatthewCroughan) [reply nixos/tests/scion/freestanding-deployment/default.nix:98]: When the services bootstrap their own PKI in a future PR, this VM test will inherit some of that functionality.

File: nixos/modules/services/networking/scion/scion-control.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (sarcasticadmin) [reply nixos/modules/services/networking/scion/scion-control.nix:61]: Going to forego this after having discussed with @MatthewCroughan

File: nixos/modules/services/networking/scion/scion-dispatcher.nix (hunk 2 above)
Comment (MatthewCroughan) [reply nixos/modules/services/networking/scion/scion-dispatcher.nix:61]: I've resolved this in the latest commit, and also added a feature to the VM test to test for robustness of restarting the dispatcher

File: nixos/modules/services/networking/scion/scion-control.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (MatthewCroughan) [reply nixos/modules/services/networking/scion/scion-control.nix:61]: Can leave this on the table for now until the next PR, and we can set up /etc any time and it won't make a difference to the user, just aesthetic.

**Review by sarcasticadmin — APPROVED** [review]
@MatthewCroughan :shipit:

**Review by sarcasticadmin — APPROVED** [review]

**Comment by Lassulus** [comment]
maybe add the test to passthru.tests on the scion package?

**Pull request merged**